ARIMA

josepablocam opened the pull request:
Here's what I have so far. Any and all feedback welcomed.

josepablocam commented:
@sryza, added conjugate gradient descent (CGD) option. So now can optimize using BOBYQA (no derivative), or with CGD.

sryza commented on src/main/scala/com/cloudera/sparkts/ARIMA.scala:
```diff
@@ -0,0 +1,566 @@
+import org.apache.commons.math3.random.RandomGenerator
+import org.apache.commons.math3.stat.regression.OLSMultipleLinearRegression
+
+
```
Add a short header comment explaining what ARIMA is and linking to the Wikipedia article

sryza commented on src/main/scala/com/cloudera/sparkts/ARIMA.scala:
```diff
@@ -0,0 +1,566 @@
+import org.apache.commons.math3.analysis.{MultivariateVectorFunction, MultivariateFunction}
+import org.apache.commons.math3.optim.nonlinear.scalar.gradient.NonLinearConjugateGradientOptimizer
+import org.apache.commons.math3.optim._
+import org.apache.commons.math3.optim.nonlinear.scalar.{ObjectiveFunctionGradient, GoalType, ObjectiveFunction}
```
Over 100 characters

sryza commented on src/main/scala/com/cloudera/sparkts/ARIMA.scala:
```diff
@@ -0,0 +1,566 @@
+   * https://www.otexts.org/fpp/8/5
+   */
+  def fitModel(
+      order:(Int, Int, Int),
```
I'd separate this into 3 params

sryza commented on src/main/scala/com/cloudera/sparkts/ARIMA.scala:
```diff
@@ -0,0 +1,566 @@
+      order:(Int, Int, Int),
+      ts: Vector[Double],
+      includeIntercept: Boolean = true,
+      method: String = "css-BOBYQA"): ARIMAModel = {
```
I'd make these strings all lower case

sryza commented on src/main/scala/com/cloudera/sparkts/ARIMA.scala:
```diff
@@ -0,0 +1,566 @@
+   * @param order The order of the model
+   * @param includeIntercept does the model include an intercept
+   * @param initParams initial parameter guesses
+   * @return
```
Add doc for the return value

sryza commented on src/main/scala/com/cloudera/sparkts/ARIMA.scala:
```diff
@@ -0,0 +1,566 @@
+  }
+
+
+  // TODO: implement MLE parameter estimates with Kalman filter.
```
I would cut this unsupported method out entirely

sryza commented on src/main/scala/com/cloudera/sparkts/ARIMA.scala:
```diff
@@ -0,0 +1,566 @@
+        val params = fitWithCSSCGD(diffedTs, order, includeIntercept, initParams)
+        new ARIMAModel(order, params, includeIntercept)
+      }
+      case "ml" => throw new UnsupportedOperationException()
```
Should throw an UnsupportedOperationException on any input that's not one of the two css methods

sryza commented on src/main/scala/com/cloudera/sparkts/ARIMA.scala:
```diff
@@ -0,0 +1,566 @@
+   * Initializes ARMA parameter estimates using the Hannan-Risannen algorithm. The process is:
+   * fit an AR(m) model of a higher order (i.e. m > max(p, q)), use this to estimate errors,
+   * then fit an OLS model of AR(p) terms and MA(q) terms. The coefficients estimated by
+   * the OLS model are returned as initial parameter estimates
```
tiny nit: period at the end of the last sentence

sryza commented on src/main/scala/com/cloudera/sparkts/ARIMA.scala:
```diff
@@ -0,0 +1,566 @@
+   */
+  private def HannanRisannenInit(
+      diffedY: Array[Double],
+      order:(Int, Int, Int),
```
Add a space after the colon

sryza commented on src/main/scala/com/cloudera/sparkts/ARIMA.scala:
```diff
@@ -0,0 +1,566 @@
+    val yTrunc = diffedY.drop(m)
+    val estimated = arTerms1.zip(
+      Array.fill(yTrunc.length)(arModel.coefficients)
+      ).map { case (v, b) => v.zip(b).map { case (yi, bi) => yi * bi}.sum + arModel.c }
```
Indent this back two spaces

sryza commented on src/main/scala/com/cloudera/sparkts/ARIMA.scala:
```diff
@@ -0,0 +1,566 @@
+  }
+
+  /**
+   * Calculate a differenced vector of a given order. Size-preserving by leaving first `order`
```
Is there already an implementation of this in `UnivariateTimeSeries`?  If not, can this go there?

josepablocam replied:
`com.cloudera.sparkts.TimeSeries` has a differences method, but this one is meant to work on `Vector`. It is also size-preserving, which is different from the one in `TimeSeries`. I've moved over to `UnivariateTimeSeries`. I agree that it makes more sense to have it there.

sryza commented on src/main/scala/com/cloudera/sparkts/ARIMA.scala:
```diff
@@ -0,0 +1,566 @@
+   * @return a new vector where the difference operation as been inverted
+   */
+  def invDifferences(ts: Vector[Double], order: Int): Vector[Double] = {
+
```
Extra space

sryza commented on src/main/scala/com/cloudera/sparkts/ARIMA.scala:
```diff
@@ -0,0 +1,566 @@
+}
+
+class ARIMAModel(
+    val order:(Int, Int, Int), // AR(p), differencing(d), MA(q)
```
Separate this into three separate arguments

josepablocam replied:
I've broken it out in the `ARIMAModel` class, do you think the same should be done in the `ARIMA` object? I have kept it as a triple for now, since that is akin to how R's stats `arima` works, which takes a `c(p, d, q)` for order.

sryza replied:
Yeah, I think it should be done in the ARIMA object as well.  I don't think we necessarily need to be consistent with R here.

sryza commented on src/main/scala/com/cloudera/sparkts/ARIMA.scala:
```diff
@@ -0,0 +1,566 @@
+    gradient := gradient :/ -sigma2
+    gradient.toArray
+  }
+
```
Extra newline

josepablocam commented:
@sryza, updated so that p,d,q terms are broken out in all functions. Some other changes:

1. improved docs a bit (and corrected given p,d,q change
2. added a new parameter to `fitModel`, a set of user provided initial parameters for the optimization
3. set default optimization to `css-cgd`, since `BOBYQA` seems to sometimes get stuck when fitting edge case `p = 0, d = 0, q = 0`
4. Added test for test case referenced in (3) 
5. small fix to Hannan-Rissanen name, I had misspelled
6. I reordered some of the function parameters in `fitWith...` and `Hannan...` to be consistent with the order in `ARIMA.fitModel`

josepablocam commented:
@sryza fixed differencing issue. Changed forecast function accordingly and created 2 functions in `UnivariateTimeSeries` to handle differencing and inverse differencing appropriately. Added test for both.

sryza commented on src/main/scala/com/cloudera/sparkts/ARIMA.scala:
```diff
@@ -0,0 +1,586 @@
+   *                       (default), initialized using Hannan-Rissanen algorithm. If provided,
+   *                       order of parameter should be: intercept term, AR parameters (in
+   *                       increasing order), MA parameters (in increasing order)
+   *
```
Extra newlines

sryza commented on src/main/scala/com/cloudera/sparkts/ARIMA.scala:
```diff
@@ -0,0 +1,586 @@
+   * @return initial ARMA(p, d, q) parameter estimates
+   */
+  private def HannanRissanenInit(
+    p: Int,
```
indent arguments an extra two spaces

sryza commented on src/main/scala/com/cloudera/sparkts/ARIMA.scala:
```diff
@@ -0,0 +1,586 @@
+   * @param newError the error at time t
+   * @return a modified array with the latest error placed into index 0
+   */
+  def updateMAErrors(errs: Array[Double], newError: Double): Unit= {
```
Need space after `Unit`

sryza commented on src/main/scala/com/cloudera/sparkts/ARIMA.scala:
```diff
@@ -0,0 +1,586 @@
+   * @return the time series resulting from the interaction of the parameters with the model's
+   *         coefficients.
+   */
+  //private
```
Why comment instead of mark this as private?

josepablocam replied:
woopsie. Flat out mistake. I had been testing something quickly in the console, so commented out private for that run. Completely forgot to uncomment. My bad

sryza commented on src/main/scala/com/cloudera/sparkts/UnivariateTimeSeries.scala:
```diff
@@ -326,5 +326,108 @@ object UnivariateTimeSeries {
+   * @param d order of differencing
+   * @return a vector of the same length differenced to order d
+   */
+  def differencesOfOrderD(ts: Vector[Double], d: Int): Vector[Double] = {
```
Would it be possible to make use of differencesAtLag in the inner function?

josepablocam replied:
yes, if we change `differencesAtLag` to take a starting index, or alternatively we can keep as it is if we are ok with the cost of splitting up the vector at index `i` into the portion that doesn't need to be differenced further (since it doesn't have enough elements prior to it) and that portion that needs to be differenced, and concatenate those after calling `differencesAtLag`. Or we do away with the recursive approach and just use a while loop. What do you think is best?

josepablocam replied:
Also just realized I should just be creating a copy of `ts` prior to calling the helper, rather than in the helper function (which means the vector gets copied with every recursive call). My bad. I will fix this once we decide on the better approach overall.

sryza replied:
differencesAtLag can take a starting index.  We can keep around an old version that calls into the new version with starting index 0.  Would that work?

josepablocam replied:
changed this, and fixed the unnecessary vector copying in `differencesOfOrderD`

josepablocam commented:
added stationarity and invertibility checks (but no transformation of parameters if they checks fail. I'm not sure we want that or not)

sryza commented:
OK, this looks ready to go in to me.  Is the WIP labeling just a relic, or are there additional changes you wanted to make?

josepablocam commented:
the WIP marker was just to note that we were still making changes. I will give it a full once-over again, and will remove after that.

josepablocam commented:
also, this patch depends on [providing AR parameter for intercept inclusion](https://github.com/cloudera/spark-timeseries/pull/32)

josepablocam commented:
@syrza removing the WIP tags, so ready to merge whenever you see fit

sryza commented:
Awesome, in that case, merging this in
